p2p/peersHolder/peersHolder.go

```diff
@@ -13,80 +14,106 @@ type peerInfo struct {
+	shardID           uint32
+	index             int
 }
 
 type peersHolder struct {
-	preferredConnAddrSlices    []string
-	connAddrSliceToPeerInfo    map[string]*peerInfo
+	preferredConnAddresses     []string
+	connAddrToPeersInfo        map[string][]*peerInfo
 	tempPeerIDsWaitingForShard map[core.PeerID]string
 	peerIDsPerShard            map[uint32][]core.PeerID
 	peerIDs                    map[core.PeerID]*peerIDData
-	sync.RWMutex
+	mut                        sync.RWMutex
 }
 
 // NewPeersHolder returns a new instance of peersHolder
-func NewPeersHolder(preferredConnectionAddressSlices [][]byte) *peersHolder {
+func NewPeersHolder(preferredConnectionAddresses []string) *peersHolder {
 	preferredConnections := make([]string, 0)
-	connAddrSliceToPeerIDs := make(map[string]*peerInfo)
+	connAddrToPeerIDs := make(map[string][]*peerInfo)
 
-	for _, connAddrSlice := range preferredConnectionAddressSlices {
-		preferredConnections = append(preferredConnections, string(connAddrSlice))
-		connAddrSliceToPeerIDs[string(connAddrSlice)] = nil
+	connectionValidator := connectionStringValidator.NewConnectionStringValidator()
+
+	for _, connAddr := range preferredConnectionAddresses {
+		if !connectionValidator.IsValid(connAddr) {
+			continue
+		}
+
+		preferredConnections = append(preferredConnections, connAddr)
+		connAddrToPeerIDs[connAddr] = nil
 	}
 
 	return &peersHolder{
-		preferredConnAddrSlices:    preferredConnections,
-		connAddrSliceToPeerInfo:    connAddrSliceToPeerIDs,
+		preferredConnAddresses:     preferredConnections,
+		connAddrToPeersInfo:        connAddrToPeerIDs,
 		tempPeerIDsWaitingForShard: make(map[core.PeerID]string),
 		peerIDsPerShard:            make(map[uint32][]core.PeerID),
 		peerIDs:                    make(map[core.PeerID]*peerIDData),
 	}
 }
 
 // PutConnectionAddress will perform the insert or the upgrade operation if the provided peerID is inside the preferred peers list
-func (ph *peersHolder) PutConnectionAddress(peerID core.PeerID, connectionAddress []byte) {
-	ph.Lock()
-	defer ph.Unlock()
+func (ph *peersHolder) PutConnectionAddress(peerID core.PeerID, connectionAddress string) {
+	ph.mut.Lock()
+	defer ph.mut.Unlock()
+
+	knownConnection := ph.getKnownConnection(connectionAddress)
+	if len(knownConnection) == 0 {
+		return
+	}
 
-	knownSlice := ph.getKnownSlice(string(connectionAddress))
-	if len(knownSlice) == 0 {
+	peersInfo := ph.connAddrToPeersInfo[knownConnection]
+	if peersInfo == nil {
+		ph.addNewPeerInfoToMaps(peerID, knownConnection)
 		return
 	}
 
-	pInfo := ph.connAddrSliceToPeerInfo[knownSlice]
+	// if we have new peer for same connection, add it to maps
+	pInfo := ph.getPeerInfoForPeerID(peerID, peersInfo)
 	if pInfo == nil {
-		ph.tempPeerIDsWaitingForShard[peerID] = knownSlice
-		ph.connAddrSliceToPeerInfo[knownSlice] = &peerInfo{
-			pid:     peerID,
-			shardID: 0, // this will be overwritten once shard is available
-		}
+		ph.addNewPeerInfoToMaps(peerID, knownConnection)
+	}
```

Comment:
missing statement as an else? what happens if pInfo != nil?

Reply:
if pInfo != nil it means that we already have that specific pid saved for a preferred connection and there is nothing to update for it